src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java

```diff
@@ -0,0 +1,302 @@
+    while (current != null) {
+      items.addFirst(current);
+      if (current == from) break;
+      current = current.getParent();
+    }
+    return items;
+  }
+
+  static public PsiFileSystemItem findRoot(PsiFileSystemItem file) {
+    //HaxelibModuleManager
+    Project project = file.getProject();
+    Module[] modules = ModuleManager.getInstance(project).getModules();
+    List<VirtualFile> roots = new ArrayList<VirtualFile>();
+    for (Module module : modules) {
+      Sdk sdk = ModuleRootManager.getInstance(module).getSdk();
+      roots.addAll(Arrays.asList(sdk.getRootProvider().getFiles(OrderRootType.CLASSES)));
+    }
+    roots.addAll(Arrays.asList(ProjectRootManager.getInstance(project).getContentSourceRoots()));
+
+    PsiFileSystemItem current = file;
+    while (current != null) {
+      for (VirtualFile root : roots) {
+        //System.out.println(root.getCanonicalPath());
+        //System.out.println(current.getVirtualFile().getCanonicalPath());
+        if (root.getCanonicalPath().equals(current.getVirtualFile().getCanonicalPath())) return current;
+      }
+      current = current.getParent();
+    }
+    return null;
+  }
+
+  static public String getDirectoryPath(PsiDirectory dir) {
+    String out = "";
+    while (dir != null) {
+      if (out.length() != 0) out = "/" + out;
+      out = dir.getName() + out;
+      dir = dir.getParent();
+    }
+    return out;
+  }
+}
+
+abstract class HaxeSemanticIntentionAction implements IntentionAction {
+  private String text;
+
+  public HaxeSemanticIntentionAction(String text) {
+    this.text = text;
+  }
+
+  @Nls
+  @NotNull
+  @Override
+  public String getText() {
+    return this.text;
+  }
+
+  @Nls
+  @NotNull
+  @Override
+  public String getFamilyName() {
+    return "semantic";
+  }
+
+  @Override
+  public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
+    return true;
+  }
+
+  @Override
+  public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
+    run();
+  }
+
+  @Override
+  public boolean startInWriteAction() {
+    return false;
+  }
+
+  abstract public void run();
+}
+
+class HaxeSemanticError {
+  public String message;
+  public IntentionAction quickfix;
+
+  public HaxeSemanticError(String message, IntentionAction quickfix) {
+    this.message = message;
+    this.quickfix = quickfix;
+  }
+
+  public HaxeSemanticError(String message) {
+    this(message, null);
+  }
+
+  static public Key<List<HaxeSemanticError>> KEY_HAXE_ERROR = new Key<List<HaxeSemanticError>>("KEY_HAXE_ERROR");
+  static public Key<Long> KEY_ANALYZED_METHOD_TIME = new Key<Long>("HAXE_ANALYZED_METHOD");
+
+  public static void clearErrors(PsiElement element) {
+    element.putUserData(KEY_HAXE_ERROR, null);
+  }
+
+  public static void addError(PsiElement element, HaxeSemanticError error) {
+    List<HaxeSemanticError> data = element.getUserData(KEY_HAXE_ERROR);
```

Comment:
I had some exceptions with that. I supposed something like that, related with multithreading and UI as happened in C# with invoke, but I didn't investigated it too much. So thanks for the info! I will change it.